Refactor to use go-mod-bootstrap package

hahattan opened the pull request:
Refactor the bootstrap and init process to use go-mod-bootstrap
Keep the globalvars at this moment because of many dependency problems in SDK
The  basic usage of dic is leveraged in `internal/client/init.go`
Also move all service-related go files (Specifically, file using `type Service struct`) inside `internal/service` dircetory. 


Fix #421

lenny-intel commented on internal/container/deviceservice.go:
```diff
@@ -0,0 +1,28 @@
+	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
+)
+
+// ConfigurationName contains the name of device service's ConfigurationStruct implementation in the DIC.
```
ConfigurationName  => DeviceServiceName

lenny-intel replied:
change `ConfigurationName` to `DeviceServiceName`

lenny-intel commented on internal/container/deviceservice.go:
```diff
@@ -0,0 +1,28 @@
+// ConfigurationName contains the name of device service's ConfigurationStruct implementation in the DIC.
+var DeviceServiceName = di.TypeInstanceToName(contract.DeviceService{})
+
+// ConfigurationFrom helper function queries the DIC and returns device service's ConfigurationStruct implementation.
```
ConfigurationFrom  => DeviceServiceFrom

lenny-intel replied:
change `ConfigurationFrom` to `DeviceServiceFrom`

lenny-intel commented on internal/controller/restrouter.go:
```diff
@@ -9,71 +9,48 @@
+	r.HandleFunc(common.APIDiscoveryRoute, discoveryFunc).Methods(http.MethodPost)
+	r.HandleFunc(common.APITransformRoute, transformFunc).Methods(http.MethodGet)
 
-func (c RestController) AddRoute(route string, handler func(http.ResponseWriter, *http.Request), methods ...string) error {
```
Can not remove this or the reservedRoutes. AddRoute is used by device-rest-go

lenny-intel commented on internal/service/service.go:
```diff
@@ -68,73 +62,6 @@ func (s *Service) AsyncReadings() bool {
-	return err
-}
-
-// AddRoute allows leveraging the existing internal webserver to add routes specific to Device Service.
```
Can not removed this API. It is used by device-rest-go.

hahattan commented:
rebased, resolved the issues pointed out by @lenny-intel (and thanks for the review)

hahattan commented:
> LGTM, should remove commented out code prior to moving out of draft.

Sure, last I want to make sure whether we can remove `initAttempts` and `initialized` field in `type Service struct` giving the fact that there's no reference on these 2 parameters in the SDK ?

hahattan commented:
Remove commented out code, move LoadRestRoutes into SDK's bootstrap handler (to be consistent with [edgex-go's latest refactoring](https://github.com/edgexfoundry/edgex-go/pull/2264))

michaelestrin commented:
> Looks good for current state, but I think it is a bit premature for this since the Registry refactoring and self populate (remove config seed) issues need to be completed and integrated into go-mod-bootstrap. These will have big impact on these changes. @michaelestrin , thoughts?

I'm impressed that you moved quickly to integrate go-mod-bootstrap.  I concur with @lenny-intel -- go-mod-bootstrap is still very much in flux (your change set specifies v0.0.1 and as of this comment we're currently at v0.0.8) and there is at least one major missing component (environment variable override support is limited to overriding the `registryInfo` value) that would cause issues if you merged this code now.

I encourage you to maintain this branch and keep updating to the latest go-mod-bootstrap version as it evolves and watch the dependent issues (https://github.com/edgexfoundry/go-mod-bootstrap/issues/14, https://github.com/edgexfoundry/go-mod-bootstrap/issues/18).

Happy to provide a code review -- please add me as a reviewer if that's desired.

cloudxxx8 commented:
thanks, @michaelestrin and @lenny-intel 
we will keep this PR opening until we feel go-mod-bootstrap is ready
for the registry and configuration refactoring, we could create a separated PR for that

brandonforster commented:
Is this still being worked?

hahattan commented:
> Is this still being worked?

Yes I'm still working on this

hahattan commented:
rebased, update to go-mod-bootstrap v0.0.18. 
One needs to be done is the SDK specific `--overwrite` argument in #383, and I think it's also dependent upon edgexfoundry/go-mod-bootstrap#14

lenny-intel commented:
There are still some pending changed to go-mod-bootstrap before it is ready for SDK (Device & App ) consumption.

hahattan commented:
rebased and updated to integrate go-mod-bootstrap v0.0.25, @lenny-intel would you please review this PR again if god-mod-bootstrap is ready for use ? 
Thanks

lenny-intel commented:
@hahattan , I fixed the issue you discovered so please switch to using v0.0.26

lenny-intel commented:
@hahattan , I have been heads down on an issue, which just got resolved, so I should be able to review this PR soon.

lenny-intel commented on internal/common/types.go:
```diff
@@ -94,10 +81,7 @@ type DeviceInfo struct {
-	EnableRemote bool
-	// File is the pathname of a local log file to be created.
-	File string
+	bootstrapConfig.LoggingInfo
```
Same issue with embedding bootstrapConfig.LoggingInfo. Remove this local version and just use bootstrapConfig.LoggingInfo directly

lenny-intel commented on internal/common/types.go:
```diff
@@ -107,7 +91,7 @@ type Config struct {
 	Service ServiceInfo
 	// Registry contains registry-specific settings.
-	Registry RegistryService
+	Registry RegistryInfo
```
Remove this whole struct and just use your new ConfigurationStruct

lenny-intel commented on internal/common/types.go:
```diff
@@ -138,14 +122,9 @@ type DeviceConfig struct {
 
 // ClientInfo provides the host and port of another service in the eco-system.
 type ClientInfo struct {
+	bootstrapConfig.ClientInfo
```
Same issue with embedding bootstrapConfig.ClientInfo. In this case the timeout is used, so you just need to create a receiver functions that returns bootstrapConfig.ClientInfo.

lenny-intel commented on pkg/service/init.go:
```diff
@@ -0,0 +1,95 @@
+func (b *Bootstrap) BootstrapHandler(ctx context.Context, wg *sync.WaitGroup, _ startup.Timer, dic *di.Container) (success bool) {
+	controller.LoadRestRoutes(b.router, dic)
+
+	common.CurrentConfig = container.ConfigurationFrom(dic.Get)
```
Please consider creating a new issue to refactor to remove using global variables.

hahattan replied:
I've tried it but couldn't come up with an efficient way to do so. 
How would you suggest do achieve this? Do we just pass the dic container all around in SDK or is there any better solution ?

brandonforster replied:
Check out the core services and how they handle this. We spent quite a bit of effort making things like the logging client, registry client, and external service clients lazy loaded and not global.

 It's certainly something that should be addressed in a follow on, because it is quite a bit of work. But, you're right, the answer is to just pass around parameters everywhere in a manner not too dissimilar from how I got the wait group into my code in my PR.

hahattan replied:
ok I'll create a new issue once this PR is merged.

lenny-intel commented on pkg/service/main.go:
```diff
@@ -0,0 +1,68 @@
+		startupTimer,
+		dic,
+		[]interfaces.BootstrapHandler{
+			// secret.NewSecret().BootstrapHandler,
```
Remove commented out code

lenny-intel commented on pkg/service/main.go:
```diff
@@ -0,0 +1,68 @@
+	startupTimer := startup.NewStartUpTimer(common.BootRetrySecondsDefault, common.BootTimeoutSecondsDefault)
+
+	if len(serviceName) == 0 || len(serviceVersion) == 0 || driver == nil {
+		_, _ = fmt.Fprintf(os.Stderr, "Please specify correct device service informations")
```
Suggest more user friendly output of which data is missing.

hahattan commented on internal/common/config.go:
```diff
@@ -0,0 +1,94 @@
+	// WritableInfo contains configuration settings that can be changed in the Registry .
+	Writable WritableInfo
+	// Clients is a map of services used by a DS.
+	Clients map[string]ClientInfo
```
Don't we still need the `Name` field while checking dependent services?
https://github.com/edgexfoundry/device-sdk-go/blob/0605716f3fd271273ac872c46dcfc3221fe9ce3f/internal/clients/init.go#L120

lenny-intel replied:
Core services don't use the `Name` field, so maybe @brandonforster's refactoring for new clients will remove the need. But for this PR, looks like you need it still, so like others, create a receiver that converts your local version to a bootstrapConfig.ClientInfo.

lenny-intel replied:
Looks like the use of constants from go-mod-core-contracts should resolve this and no long need the `Name` field. 
https://github.com/edgexfoundry/go-mod-core-contracts/blob/master/clients/constants.go#L61

lenny-intel commented on internal/clients/init.go:
```diff
@@ -181,22 +162,22 @@ func checkConsulAvailable() bool {
 	// initialize Core Metadata clients
-	params.ServiceKey = common.CurrentConfig.Clients[common.ClientMetadata].Name
+	// hard-coded here, should be fixed in PR#447
+	params.ServiceKey = "edgex-core-metadata"
```
These constants are available from go-mod-core-contracts. Please use them and then PR#447 doesn't have to fix this.
https://github.com/edgexfoundry/go-mod-core-contracts/blob/master/clients/constants.go#L61

brandonforster replied:
This is going to be immediately overwritten anyway so I wouldn't worry about it.

lenny-intel commented on internal/common/config.go:
```diff
@@ -0,0 +1,90 @@
+// structure -- until we can make backwards-breaking configuration.toml changes (which would consolidate these fields
+// into an interfaces.BootstrapConfiguration struct contained within ConfigurationStruct).
+func (c *ConfigurationStruct) GetBootstrap() interfaces.BootstrapConfiguration {
+	clients := make(map[string]bootstrapConfig.ClientInfo)
+	for k, v := range c.Clients {
+		clients[k] = v.GetBootstrapClientInfo()
+	}
```
Per the comments above you should now be able to just use `bootstrapConfig.ClientInfo` directly.

lenny-intel commented on internal/common/types.go:
```diff
@@ -138,19 +106,41 @@ type DeviceConfig struct {
 
 // ClientInfo provides the host and port of another service in the eco-system.
 type ClientInfo struct {
```
Can remove now that `Name` field isn't needed.

brandonforster commented:
Any chance we could merge this? The sooner I have final code in master the sooner I can open my PR for review.
